wallet-new/src/Cardano/Wallet/Kernel/Restore.hs

```diff
@@ -0,0 +1,193 @@
+              -> (Blund -> IO (Map HD.HdAccountId PrefilteredBlock, [TxMeta]))
+              -> IO Coin
+restoreWallet wallet hdRoot (wId, wdc) prefilter = do
+    -- Find all of the current UTXO that this wallet owns.
+    mine <- withNodeState (wallet ^. walletNode) (\_lock -> filterMyUtxo)
+
+    -- Create the wallet and populate the accounts we've seen.
+    let db = wallet ^. wallets
+        utxoByAcct = map (\(u, a, wam) ->
+                                 let hdAddr = initHdAddress (toHdAddressId wId wam) (InDb a)
+                                     hdAddrId = hdAddr ^. HD.hdAddressId
+                                 in (hdAddr ^. HD.hdAddressAccountId, (u, [(hdAddrId, a)]))) mine
+
+    csr <- getChainStateRestoration (wallet ^. walletNode) NotYetLocked
+    whenJust (csrCurrent csr) $ \brief -> do
+        -- Create the wallet
+        let tgtTip  = cbTip brief
+            tgtSlot = cbSlotId brief
+        void $ update db (RestoreHdWallet hdRoot tgtSlot (M.fromListWith (<>) utxoByAcct))
+
+        -- Begin restoring the wallet history in the background.
+        restoreTask <- async (restoreWalletHistoryAsync wallet wId tgtTip tgtSlot prefilter)
+        link restoreTask
+
+        -- Set the wallet's restoration information
+        slotCount <- getSlotCount (wallet ^. walletNode)
+        let restoreInfo = WalletRestorationInfo
+              { _wriCurrentSlot = 1 -- FIXME: this is temporary, should be 0
+              , _wriTargetSlot  = flattenSlotIdExplicit slotCount (cbSlotId brief)
+              , _wriThroughput  = MeasuredIn 0
+              , _wriCancel      = cancel restoreTask
+              }
+        modifyMVar_ (wallet ^. walletRestorationTask) (pure . M.insert wId restoreInfo)
+
+    -- Return the wallet's current balance.
+    let myUtxo  = foldr (\(u,_,_) -> M.union u) M.empty mine
+    return (getTotalCoinsInUtxo myUtxo)
+
+  where
+    txoAddr = txOutAddress . toaOut . snd
+
+    filterMyUtxo :: (MonadDBRead m, MonadUnliftIO m) => m [(Utxo, Address, WAddressMeta)]
+    filterMyUtxo = runConduitRes
+         $ mapOutputMaybe (\u -> let a = txoAddr u
+                                     wam = decryptAddress wdc a
+                                 in (M.fromList [u],a,) <$> wam) utxoSource
+        .| Conduit.fold (flip (:)) []
+
+-- | Restore a wallet's transaction history.
+restoreWalletHistoryAsync :: Kernel.PassiveWallet
+                          -> WalletId
+                          -> HeaderHash
+                          -> SlotId
+                          -> (Blund -> IO (Map HD.HdAccountId PrefilteredBlock, [TxMeta]))
+                          -> IO ()
+restoreWalletHistoryAsync wallet wId target tgtSlot prefilter = do
+
+    say "sleeping 20 seconds before starting the restoration, so that the tip has a chance to move"
+    threadDelay 20000000
+
+    -- Get the genesis block.
+    genesisBlock <- withNodeState (wallet ^. walletNode) $ \_lock ->
+        getBlock =<< getFirstGenesisBlockHash
```

Comment:
Oh, nice catch! I _think_ that `resolveForwardLink` always gives you `MainBlock`s; this is important only because the restoration task needs `Blund`s for prefiltering, and as far as I understand it we can only get `Blund`s for `MainBlock`s (but I can't recall why I believe that)